### PR TITLE
GPT integration

### PR DESCRIPTION
**Pull request by Cephian**
Integration of OpenAI API

### CODE REVIEW

File: src/gpt.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    static_context = context_data["static_context"]
+    author_context = await get_author_context(message)
+    optional_context = get_optional_context(content)
+    context = "\n".join(static_context + author_context + optional_context)
+
+    token_allowance = config.GPT_REQUEST_TOKEN_LIMIT - token_count(context)
+    history = await fetch_history(token_allowance, message)
+    # We couldn't git even a single message in history
+    if not history:
+        message.reply("I'm not reading all that.")
```
Comment (anoadragon453) [inline src/gpt.py:142]: Looks like a missing return here.

File: src/gpt.py [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+# Build context around a message
+async def get_message_context(message: Message) -> str:
+    # build contexts
+    static_context = context_data["static_context"]
+    author_context = await get_server_context(message)
+    random_context = get_random_context()
+    return static_context + author_context + random_context
+
+
+# Query thge OpenAI API and return response
```
Comment (anoadragon453) [inline src/gpt.py:214]:
```suggestion
# Query the OpenAI API and return response
```

File: src/gpt.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+    response = await query_api(data)
+    if response:
+        # Remove "Busty: " prefix
+        prefix = f"{get_name(self_user)}: "
+        if response.startswith(prefix):
+            response = response[len(prefix) :]
+
+        return response
+    else:
+        await message.reply("busy rn")
```
Comment (anoadragon453) [inline src/gpt.py:284]: As the other block will just exit.

```suggestion

    await message.reply("busy rn")
```

File: src/gpt.py [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+    async with gpt_lock:
+        async with message.channel.typing():
+            response = await get_response_text(message)
+        if response:
+            response_split = [
+                response[i : i + config.MESSAGE_LIMIT]
+                for i in range(0, len(response), config.MESSAGE_LIMIT)
+            ]
+            # If user's message is most recent message in channel, send
+            # If others have been sent in the meantime, reply
```
Comment (anoadragon453) [inline src/gpt.py:304]: Just to make this read a bit clearer (ChatGPT was not used to generate this diff):

```suggestion
            # Reply to the message if it's not the most recent on in the chat history
```

File: src/gpt.py [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+            most_recent_message = [
+                msg async for msg in message.channel.history(limit=1)
+            ][0]
+            for idx, text in enumerate(response_split):
+                if idx == 0 and message == most_recent_message:
+                    await message.channel.send(response)
+                else:
+                    await message.reply(response)
```
Comment (anoadragon453) [inline src/gpt.py:312]: Currently this code will turn any subsequent messages that were split up due to message length into replies. Is that what you meant?

I would have assumed that in the case of needing a reply, only the first message would be a reply, and all subsequent ones would not be.

File: src/main.py [hunk 6]
```diff
@@ -44,6 +55,21 @@ async def on_close() -> None:
...
 
 
+@client.event
+async def on_message(message: Message) -> None:
+    if (
+        config.openai_api_key
+        and message.guild
+        and (
+            client.user in message.mentions
+            or any(role.name == client.user.name for role in message.role_mentions)
```
Comment (anoadragon453) [inline src/main.py:65]:
```suggestion
            # For the case where a user accidentally mentions the bot's role
            # instead of their nick (which are typically named the same).
            or any(role.name == client.user.name for role in message.role_mentions)
```

File: src/gpt.py [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+        if response.startswith(prefix):
+            response = response[len(prefix) :]
+
+        return response
+    else:
+        await message.reply("busy rn")
+    return None
+
+
+async def respond(message: Message):
```
Comment (anoadragon453) [inline src/gpt.py:288]:
```suggestion
async def respond(message: Message) -> None:
```

File: src/gpt.py [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+            timeout=10.0,
+        )
+        return response["choices"][0]["message"]["content"]
+
+    except Exception as e:
+        print("OpenAI API exception:", e)
+        return None
+
+
+def get_history_context(history: List[Tuple[str, bool]]):
```
Comment (anoadragon453) [inline src/gpt.py:229]:
```suggestion
def get_history_context(history: List[Tuple[str, bool]]) -> List[str]:
```

File: src/gpt.py [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+def get_name(user):
+    user_info = context_data["user_info"]
+    id = str(user.id)
+    if id in user_info and "name" in user_info[id]:
+        return user_info[id]["name"]
+    return user.name
+
+
+# Get totally random triggers about how to act:
+def get_random_context():
```
Comment (anoadragon453) [inline src/gpt.py:90]:
```suggestion
def get_random_context() -> List[str]:
```

File: src/gpt.py [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+
+
+# Check if a message should not be allow in API history
+# Currently this is just if it contains banned phrases
+def disallowed_message(message: Message) -> bool:
+    return banned_word_pattern.search(message.content.lower()) is not None
+
+
+# Get the name we should call the user
+def get_name(user):
```
Comment (anoadragon453) [inline src/gpt.py:81]:
```suggestion
def get_name(user: Member) -> str:
```

File: src/gpt.py [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+
+import openai
+import tiktoken
+from nextcord import Member, Message
+
+import config
+
+
+# Initialize globals
+def initialize(client):
```
Comment (anoadragon453) [inline src/gpt.py:16]:
```suggestion
def initialize(client: Client) -> None:
```

**Review by anoadragon453 — CHANGES_REQUESTED** [review]
Architecturally this looks pretty good overall. A bunch of little fixes and suggestions, but nothing fundamentally needs changing.

File: src/gpt.py [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+    return banned_word_pattern.search(message.content.lower()) is not None
+
+
+# Get the name we should call the user
+def get_name(user):
+    user_info = context_data["user_info"]
+    id = str(user.id)
+    if id in user_info and "name" in user_info[id]:
+        return user_info[id]["name"]
+    return user.name
```
[the remark this answers is not in the document]
Comment (Cephian) [reply src/gpt.py:86]: I chose Discord name instead of nick because people seem to have constantly changing nicknames in the server that we don't call them. Just now kley is "kylie", fin is "“Real” Russian Bank Assistant", ember is "embird", emi is "emillion", kat is "kat burglar", etc. Yet in convos we still tend to refer to them in convo by their "true" name. Some of this is fixed by the hardcoded nicknames list but I think there's a much higher "true name" hit rate by nick and not by whatever joke they they currently have.

File: src/gpt.py [hunk 13]
```diff
@@ -0,0 +1,312 @@
...
+                [
+                    user["name"].lower()
+                    for user in context_data["user_info"].values()
+                    if "info" in user
+                ]
+            )
+        )
+    )
+    # Store map for user info triggers
+    user_info_map = {
```
[the remark this answers is not in the document]
Comment (Cephian) [reply src/gpt.py:67]: ¯\\_(ツ)_/¯

my reasoning is that even if it may be slightly less efficient for our small size, it's still currently dwarfed by API response time, definitely still /pretty/ fast, and it will scale much better so we don't have to re-test if we add more things.

File: src/gpt.py [hunk 14]
```diff
@@ -0,0 +1,312 @@
...
+    token_limit: int, speaking_turn_limit: int, message: Message
+) -> List[Tuple[str, bool]]:
+    total_tokens = 0
+    history = []
+    idx = 0
+    one_hour = datetime.timedelta(hours=1)
+    seen_message = False
+    speaking_turn_count = 0
+    last_author = None
+    async for msg in message.channel.history():
```
[the remark this answers is not in the document]
Comment (Cephian) [reply src/gpt.py:169]: I also thought I could just do that, but enumerate doesn't work with async iterators. There are [third-party implementations](https://gist.github.com/tebeka/8c6b0589f5783bc4115a) but as far as I can tell there's nothing in standard libraries.

File: src/gpt.py [hunk 15]
```diff
@@ -0,0 +1,312 @@
...
+        # Break if maximum number of speaking turns has been reached
+        if msg.author != last_author:
+            speaking_turn_count += 1
+            if speaking_turn_count > speaking_turn_limit:
+                break
+            last_author = msg.author
+
+        # Break if message is disallowed
+        if disallowed_message(msg):
+            break
```
[the remark this answers is not in the document]
Comment (Cephian) [reply src/gpt.py:192]: I was thinking that 
1. if these terms are being thrown around Busty might be in "LLM Mode" and need a full reset
2. Context might not "make sense" with stuff missing, though i think chatgpt is not really trained to call out no-sense convos so this might not be an issue

I'd be ok with just skipping

File: src/gpt.py [hunk 16]
```diff
@@ -0,0 +1,312 @@
...
+            response = response[len(prefix) :]
+
+        return response
+    else:
+        await message.reply("busy rn")
+    return None
+
+
+async def respond(message: Message):
+    # React with a "stop" hand, we're responding to someone else
```
[the remark this answers is not in the document]
Comment (Cephian) [reply src/gpt.py:291]: Yeah the current implementation is a little sketchy, but I think good enough since 99% of the time in Busty's people congregate in one channel. It's definitely not multiserver or multichannel friendly though.

File: src/gpt.py [hunk 17]
```diff
@@ -0,0 +1,312 @@
...
+    global user_trigger_pattern
+    global user_info_map
+
+    openai.api_key = config.openai_api_key
+    # Global lock for message response
+    gpt_lock = asyncio.Lock()
+    # Load manual hidden data
+    try:
+        with open(config.gpt_context_file) as f:
+            context_data = json.load(f)
```
[the remark this answers is not in the document]
Comment (Cephian) [reply src/gpt.py:32]: Yeah, I think we should make an issue for it and keep it temporarily undocumented

File: src/gpt.py [hunk 18]
```diff
@@ -0,0 +1,312 @@
...
+        context.append("Write long detailed responses")
+    return context
+
+
+# Get context about the server
+async def get_server_context(message: Message) -> str:
+    result = []
+    # Load server event info
+    if message.guild and message.guild.scheduled_events:
+        # No idea if these are sorted by time
```
[the remark this answers is not in the document]
Comment (Cephian) [reply src/gpt.py:107]: I think my preferred solution would be to always take the soonest event, then if there's no ":" splitting character we say

"next event" instead of "next bust event" or something

File: src/gpt.py [hunk 19]
```diff
@@ -0,0 +1,312 @@
...
+    idx = 0
+    one_hour = datetime.timedelta(hours=1)
+    seen_message = False
+    speaking_turn_count = 0
+    last_author = None
+    async for msg in message.channel.history():
+        # Skip messages newer than what we're replying to
+        seen_message = seen_message or (msg == message)
+        if not seen_message:
+            continue
```
[the remark this answers is not in the document]
Comment (Cephian) [reply src/gpt.py:175]: Trying this and it seems a little strange. The API makes it sound like you can pass in either an `int` id or a `datetime`, but actually you pass in an object which has an "id" attribute (aka the message itself) or a datetime. Using `before=message` doesn't quite work for our code flow since it doesn't include `message` itself. I tried using `before=message.created_at+datetime.timedelta(microseconds=1)` but even that seems to not work so well. In fact even with a timedelta of one second it seems to not work?

There might be a way to "zip" together [message] + message.channel.history(before=message) in an async iterator friendly way (aka without flattening in advance) but I'm not sure if it would add too much complication to be worth it

File: src/gpt.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (anoadragon453) [reply src/gpt.py:171]: Alas, that would do it.

File: src/gpt.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (anoadragon453) [reply src/gpt.py:175]: Oh yeah, I had misread [`nextcord.abc.Snowflake`](https://docs.nextcord.dev/en/stable/api.html#nextcord.abc.Snowflake) as an alias of `int`, rather than having an `int` attribute.

Indeed, it looks like it's going to be more complicated to reason about prepending to nextcord's `AsyncIterator` than just skipping via this way, especially since 99% of the time we'll be responding to the latest message in history anyways.

File: src/gpt.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (anoadragon453) [reply src/gpt.py:194]: I suggest trying it briefly and seeing if it helps. Most of this is experimentation anyhow :art:

[the remark this answers is not in the document]
Comment (anoadragon453) [reply src/gpt.py:194]: I suggest trying it briefly and seeing if it helps. Most of this is experimentation anyways :art:

File: src/gpt.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (anoadragon453) [reply src/gpt.py:107]: Oh, to give the bot context on any event? Indeed, that could be interesting!

File: src/gpt.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (anoadragon453) [reply src/gpt.py:67]: Oops, I meant to address that question myself before submitting the review haha.

It looked like a slightly odd data structure on first glance. But looking again with fresh eyes, it looks fine.

File: src/gpt.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (anoadragon453) [reply src/gpt.py:32]: Cool, can you do that?

File: src/gpt.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (anoadragon453) [reply src/gpt.py:86]: That's fair. On the contrary Evershade's discord username is "sillyzippy" which typically nobody calls them. And it may miss out on fun opportunities for people to change their name to "Obama" or whatnot and start talking with the bot. But it'll have to be something we gauge in practice I think.

For now let's leave it as it is. We should probably strip out the `#1234` that is appended to the end of Discord usernames though?

[the remark this answers is not in the document]
Comment (Cephian) [reply src/gpt.py:86]: > We should probably strip out the #1234 that is appended to the end of Discord usernames though?

`user.name` already has the discriminator stripped

File: src/gpt.py (hunk 5 above)
Comment (Cephian) [reply src/gpt.py:313]: Yeah I messed up the logic, fixed

File: src/gpt.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (Cephian) [reply src/gpt.py]: Added

File: src/gpt.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (Cephian) [reply src/gpt.py]: I tried a number of things and while I got her to admit to being an LLM a few times, I definitely used some knowledge of specific trigger words to get around and stuff. I think just skipping bad messages instead of breaking on them is good enough.

image

File: src/gpt.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (Cephian) [reply src/gpt.py]: #184

File: src/llm.py [hunk 20]
```diff
@@ -0,0 +1,317 @@
...
+
+
+# Fetch message content from history up to a certain token allowance
+async def fetch_history(
+    token_limit: int, speaking_turn_limit: int, message: Message
+) -> List[Tuple[str, bool]]:
+    total_tokens = 0
+    # A list of (str, bool) tuples containing:
+    #     - "{message author}: {message content}"
+    #     - True if the the message was sent by the bot user, False otherwise
```
Comment (anoadragon453) [inline src/llm.py:168]:
```suggestion
    #     - True if the message was sent by the bot user, False otherwise
```

File: src/llm.py [hunk 21]
```diff
@@ -0,0 +1,317 @@
...
+
+        # Don't include messages which are both more than 3 back and an hour old
+        # as the discussion has likely moved on.
+        idx += 1
+        timezone = message.created_at.tzinfo
+        now = datetime.datetime.now(timezone)
+        if idx > 3 and (now - message.created_at) > one_hour:
+            break
+
+        # Break if maximum number of speaking turns has been reached
```
Comment (anoadragon453) [inline src/llm.py:193]:
```suggestion
        # Break if maximum number of speaking turns has been reached
        # We use turns instead of a simple message count as some people like to
        # break up their thoughts into many separate messages.
```

File: src/llm.py [hunk 22]
```diff
@@ -0,0 +1,317 @@
...
+    gpt_lock = asyncio.Lock()
+    # Load manual hidden data
+    try:
+        with open(config.llm_context_file) as f:
+            context_data = json.load(f)
+    except (FileNotFoundError, json.decoder.JSONDecodeError) as e:
+        print(
+            f"ERROR: Issue loading {config.llm_context_file}. GPT capabilities will be disabled.\n{e}"
+        )
+        # Signal context data was not loaded correctly
```
Comment (anoadragon453) [inline src/llm.py:37]: This comment seems a bit superfluous with the print statement above. Might just want to axe it?

File: src/llm.py [hunk 23]
```diff
@@ -0,0 +1,317 @@
...
+            "Be succinct",
+            "Don't be too wordy",
+            "Write long detailed responses",
+        ]
+    )
+    return [random_context]
+
+
+# Get context about the server
+async def get_server_context(message: Message) -> str:
```
Comment (anoadragon453) [inline src/llm.py:103]:
```suggestion
async def get_server_context(message: Message) -> List[str]:
```

File: src/llm.py [hunk 24]
```diff
@@ -0,0 +1,317 @@
...
+    if random.random() < 0.05 or disallowed_message(message):
+        # If message is disallowed (or the user is unlucky), pass a special instruction
+
+        # Get user info since we're not passing history which would trigger it
+        user = get_name(message.author)
+        context.append(f"{user}: {user_info_map[user.lower()]}")
+        # Pass special instruction
+        context.append(context_data["banned_phrase_instruction"])
+        # Make up empty history so she understands the format
+        history = [(f"{self_user.name}:", True), (f"{user}:", False)]
```
Comment (anoadragon453) [inline src/llm.py:262]:
```suggestion
        # Make up empty history so the bot understands how to format the response
        history = [(f"{self_user.name}:", True), (f"{user}:", False)]
```

File: src/llm.py [hunk 25]
```diff
@@ -0,0 +1,317 @@
...
+    # Respond to message
+    async with gpt_lock:
+        async with message.channel.typing():
+            response = await get_response_text(message)
+        if response:
+            response_split = [
+                response[i : i + config.MESSAGE_LIMIT]
+                for i in range(0, len(response), config.MESSAGE_LIMIT)
+            ]
+            # Reply to the message if it's not the most recent on in the chat history
```
Comment (anoadragon453) [inline src/llm.py:309]:
```suggestion
            # Reply to the message if it's not the most recent one in the chat history
```

File: src/main.py [hunk 26]
```diff
@@ -44,6 +55,23 @@ async def on_close() -> None:
...
+async def on_message(message: Message) -> None:
+    if (
+        config.openai_api_key
+        and message.guild
+        and (
+            client.user in message.mentions
+            # For the case where a user accidentally mentions the bot's role
+            # instead of their nick (which are typically named the same).
+            or any(role.name == client.user.name for role in message.role_mentions)
+            or random.random() < 1 / 150
```
Comment (anoadragon453) [inline src/main.py:68]:
```suggestion
            # Randomly respond to some messages, even if the bot is not mentioned.
            or random.random() < 1 / 150
```

File: src/main.py [hunk 27]
```diff
@@ -44,6 +55,23 @@ async def on_close() -> None:
         await bc.finish(say_goodbye=False)
 
 
+@client.event
+async def on_message(message: Message) -> None:
+    if (
+        config.openai_api_key
+        and message.guild
```
Comment (anoadragon453) [inline src/main.py:62]:
```suggestion
        # Ignore DMs to the bot.
        and message.guild
```

**Review by anoadragon453 — APPROVED** [review]
Just lots of small suggestions, but otherwise LGTM!
